### ISSUE TITLE
[run-clang-tidy.py] Add --config-file=<string>  option

### ISSUE THREAD

**Issue by bingozhu0197** [llvm/llvm-project#53745]
As clang-tidy support "
  --config-file=<string>                   -
                                             Specify the path of .clang-tidy or custom config file:
". 
Could we also add this to the script? There are cases which we want to use config files.

**Comment by SAtacker** [comment]
Hello, 
I would like to work on this. 
Thanks.

**Comment by EugeneZelenko** [comment]
Just assign yourself.

**Comment by SAtacker** [comment]
I think the `--config-file=<string>`  option is already documented [here](https://clang.llvm.org/extra/clang-tidy/#:~:text=clang%2Dtidy%20for%20each%20source,after%20reading%20specified%20config%20file.)

**Comment by JonasToth** [comment]
for reference: https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py#L226

this is the place for the argument parsing of the old `config` option in the run-clang-tidy script.
It can be expanded directly there, i think.

**Comment by SAtacker** [comment]
> for reference: https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py#L226
> 
> this is the place for the argument parsing of the old `config` option in the run-clang-tidy script. It can be expanded directly there, i think.

Thanks, I now understand the reason.

<details>
<summary>Background study</summary>
I searched a bit and found.
<pre>
satacker@ubuntu:/home/extended/llvm-project/clang-tools-extra/clang-tidy$ grep -rnw . -e '--config-file'
./tool/ClangTidyMain.cpp:183: e.g. --config-file=/some/path/myTidyConfigFile
./tool/ClangTidyMain.cpp:186:Use either --config-file or --config, not both.
./tool/ClangTidyMain.cpp:342:      llvm::errs() << "Error: --config-file and --config are "
</pre>
Which is used as a target and the script is different alltogether.
<pre>
satacker@ubuntu:/home/extended/llvm-project/clang-tools-extra/clang-tidy$ grep -rnw . -e 'ClangTidyMain.cpp'
./tool/ClangTidyMain.cpp:1://===--- tools/extra/clang-tidy/ClangTidyMain.cpp - Clang tidy tool -------===//
./tool/CMakeLists.txt:10:set(LLVM_OPTIONAL_SOURCES ClangTidyMain.cpp ClangTidyToolMain.cpp)
./tool/CMakeLists.txt:13:  ClangTidyMain.cpp
</pre>
</details>

**Comment by SAtacker** [comment]
Thanks,
You can find my commit [here](https://reviews.llvm.org/D120331).

**Comment by SAtacker** [comment]
Gentle reminder https://reviews.llvm.org/D120331
CC @JonasToth 
Thank You

**Comment by SAtacker** [comment]
Made the required changes. Thanks @JonasToth

**Comment by SAtacker** [comment]
@JonasToth Does the `Accepted` label on the review means the same as committing to the project? If not could you please commit on my behalf? Thanks for this.

**Comment by EugeneZelenko** [comment]
No, somebody must commit your changes. It'll be better to ping on review, so if Jonas is busy, somebody else could do this.

**Comment by SAtacker** [comment]
> No, somebody must commit your changes. It'll be better to ping on review, so if Jonas is busy, somebody else could do this.

Okay, thanks.

**Comment by danielbisar** [comment]
I think this change introduces a bug. run-clang-tidy-15 currently is not working. It exists with the error message: Unable to run clang-tidy. 

I am running Ubuntu 20.04 LTS, with the official apt.llvm.org repos. Current version: clang-tidy-15/unknown 1:15~++20220405074512+72fe439a4e11-1~exp1~20220405074524.211 amd64

This occurs in line 305ff and exception is handled in line 319. Adding a except Exception as e and print out the exception reveals the actual error message: ''Namespace' object has no attribute 'config_path'\ Which seems to make sense since the configuration option is called config-file not config-path. So it should be named args.config_file

There seem to be at least to spots where this needs to be corrected. The output of -list-chacks is still wrong.

**Comment by AaronBallman** [comment]
@danielbisar -- does https://github.com/llvm/llvm-project/commit/04b42c99f62216263662d1033fa9b3e12c65563e resolve the issue for you?

**Comment by danielbisar** [comment]
@AaronBallman the first part yes. The part that the listed checks are not the ones used not.

**Comment by AaronBallman** [comment]
@danielbisar thanks for the details!
@aaronmondal, would you be interested in tackling the listed checks issue as well?

**Comment by aaronmondal** [comment]
@AaronBallman Thanks for mentioning me. I'll take a look at this.

@danielbisar I'll try to reproduce. The `.clang-tidy` config file you are using may actually be relevant here as well. I vaguely remember an issue where the `.clang-tidy` config option
```
readability-identifier-naming.TemplateParameterCase
```
led to an issue which may be similar to the one you are describing.

**Comment by aaronmondal** [comment]
@danielbisar Ok I just tried to reproduce (custom build from ed4e6e03981a37d54c4ac635b2bdb2c2fe8fec92).
What i said about `TemplateParameterCase` is irrelevant here, so ignore that.

I was not able to repro the issue. Could it be that there is a typo in your `.clang-tidy` file? I recall that the recommended syntax for the `Checks` section was recently ([discussion](https://github.com/llvm/llvm-project/issues/53737), [release notes](https://github.com/llvm/llvm-project/commit/ef19de52ed59a739b0381f2b9b41604e7fa49b59)) changed to
```
Checks: |
  -*
  bugprone-*
  google-*
```
Before that change `clang-tidy` would silently ignore multiline `Checks` after a missing comma.

**Comment by danielbisar** [comment]
Hey, good point, I indeed found missing commas. Will check that further and give you an update.

**Comment by danielbisar** [comment]
@aaronmondal Okay I tried it with Version '1:15~++20220406052928+560856d56ad9-1~exp1~20220406173025.213' 
+ run-clang-tidy-15 replace with the version from https://github.com/llvm/llvm-project/commit/04b42c99f62216263662d1033fa9b3e12c65563e

Folder structure:

```
ws/src/project/
                 - .clang-tidy
                 - subfolder/code.cpp
ws/build/compile_commands.json
```
Execution tried with the following:


```
ws/src/project $ run-clang-tidy-15 -p ../../build subfolder/code.cpp
ws/src/project $ run-clang-tidy-15 -p ../../build subfolder/code.cpp
```
.clang-tidy (a stripped down version, just for this test)

```
---
Checks: "
  llvm-header-guard,
  modernize-use-nullptr,
  modernize-use-using"
WarningsAsErrors: ""
AnalyzeTemporaryDtors: false
FormatStyle: none

---


```

Output

```
Enabled checks:
    clang-analyzer-apiModeling.Errno
    clang-analyzer-apiModeling.StdCLibraryFunctions
    clang-analyzer-apiModeling.TrustNonnull
    clang-analyzer-apiModeling.TrustReturnsNonnull
    clang-analyzer-apiModeling.google.GTest
    clang-analyzer-apiModeling.llvm.CastValue
    clang-analyzer-apiModeling.llvm.ReturnValue
    clang-analyzer-core.CallAndMessage
    clang-analyzer-core.CallAndMessageModeling
    clang-analyzer-core.DivideZero
    clang-analyzer-core.DynamicTypePropagation
    clang-analyzer-core.NonNullParamChecker
    clang-analyzer-core.NonnilStringConstants
    clang-analyzer-core.NullDereference
    clang-analyzer-core.StackAddrEscapeBase
    clang-analyzer-core.StackAddressEscape
    clang-analyzer-core.UndefinedBinaryOperatorResult
    clang-analyzer-core.VLASize
    clang-analyzer-core.builtin.BuiltinFunctions
    clang-analyzer-core.builtin.NoReturnFunctions
    clang-analyzer-core.uninitialized.ArraySubscript
    clang-analyzer-core.uninitialized.Assign
    clang-analyzer-core.uninitialized.Branch
    clang-analyzer-core.uninitialized.CapturedBlockVariable
    clang-analyzer-core.uninitialized.UndefReturn
    clang-analyzer-cplusplus.InnerPointer
    clang-analyzer-cplusplus.Move
    clang-analyzer-cplusplus.NewDelete
    clang-analyzer-cplusplus.NewDeleteLeaks
    clang-analyzer-cplusplus.PlacementNew
    clang-analyzer-cplusplus.PureVirtualCall
    clang-analyzer-cplusplus.SelfAssignment
    clang-analyzer-cplusplus.SmartPtrModeling
    clang-analyzer-cplusplus.StringChecker
    clang-analyzer-cplusplus.VirtualCallModeling
    clang-analyzer-deadcode.DeadStores
    clang-analyzer-fuchsia.HandleChecker
    clang-analyzer-nullability.NullPassedToNonnull
    clang-analyzer-nullability.NullReturnedFromNonnull
    clang-analyzer-nullability.NullabilityBase
    clang-analyzer-nullability.NullableDereferenced
    clang-analyzer-nullability.NullablePassedToNonnull
    clang-analyzer-nullability.NullableReturnedFromNonnull
    clang-analyzer-optin.cplusplus.UninitializedObject
    clang-analyzer-optin.cplusplus.VirtualCall
    clang-analyzer-optin.mpi.MPI-Checker
    clang-analyzer-optin.osx.OSObjectCStyleCast
    clang-analyzer-optin.osx.cocoa.localizability.EmptyLocalizationContextChecker
    clang-analyzer-optin.osx.cocoa.localizability.NonLocalizedStringChecker
    clang-analyzer-optin.performance.GCDAntipattern
    clang-analyzer-optin.performance.Padding
    clang-analyzer-optin.portability.UnixAPI
    clang-analyzer-osx.API
    clang-analyzer-osx.MIG
    clang-analyzer-osx.NSOrCFErrorDerefChecker
    clang-analyzer-osx.NumberObjectConversion
    clang-analyzer-osx.OSObjectRetainCount
    clang-analyzer-osx.ObjCProperty
    clang-analyzer-osx.SecKeychainAPI
    clang-analyzer-osx.cocoa.AtSync
    clang-analyzer-osx.cocoa.AutoreleaseWrite
    clang-analyzer-osx.cocoa.ClassRelease
    clang-analyzer-osx.cocoa.Dealloc
    clang-analyzer-osx.cocoa.IncompatibleMethodTypes
    clang-analyzer-osx.cocoa.Loops
    clang-analyzer-osx.cocoa.MissingSuperCall
    clang-analyzer-osx.cocoa.NSAutoreleasePool
    clang-analyzer-osx.cocoa.NSError
    clang-analyzer-osx.cocoa.NilArg
    clang-analyzer-osx.cocoa.NonNilReturnValue
    clang-analyzer-osx.cocoa.ObjCGenerics
    clang-analyzer-osx.cocoa.RetainCount
    clang-analyzer-osx.cocoa.RetainCountBase
    clang-analyzer-osx.cocoa.RunLoopAutoreleaseLeak
    clang-analyzer-osx.cocoa.SelfInit
    clang-analyzer-osx.cocoa.SuperDealloc
    clang-analyzer-osx.cocoa.UnusedIvars
    clang-analyzer-osx.cocoa.VariadicMethodTypes
    clang-analyzer-osx.coreFoundation.CFError
    clang-analyzer-osx.coreFoundation.CFNumber
    clang-analyzer-osx.coreFoundation.CFRetainRelease
    clang-analyzer-osx.coreFoundation.containers.OutOfBounds
    clang-analyzer-osx.coreFoundation.containers.PointerSizedValues
    clang-analyzer-security.FloatLoopCounter
    clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling
    clang-analyzer-security.insecureAPI.SecuritySyntaxChecker
    clang-analyzer-security.insecureAPI.UncheckedReturn
    clang-analyzer-security.insecureAPI.bcmp
    clang-analyzer-security.insecureAPI.bcopy
    clang-analyzer-security.insecureAPI.bzero
    clang-analyzer-security.insecureAPI.decodeValueOfObjCType
    clang-analyzer-security.insecureAPI.getpw
    clang-analyzer-security.insecureAPI.gets
    clang-analyzer-security.insecureAPI.mkstemp
    clang-analyzer-security.insecureAPI.mktemp
    clang-analyzer-security.insecureAPI.rand
    clang-analyzer-security.insecureAPI.strcpy
    clang-analyzer-security.insecureAPI.vfork
    clang-analyzer-unix.API
    clang-analyzer-unix.DynamicMemoryModeling
    clang-analyzer-unix.Malloc
    clang-analyzer-unix.MallocSizeof
    clang-analyzer-unix.MismatchedDeallocator
    clang-analyzer-unix.Vfork
    clang-analyzer-unix.cstring.BadSizeArg
    clang-analyzer-unix.cstring.CStringModeling
    clang-analyzer-unix.cstring.NullArg
    clang-analyzer-valist.CopyToSelf
    clang-analyzer-valist.Uninitialized
    clang-analyzer-valist.Unterminated
    clang-analyzer-valist.ValistBase
    clang-analyzer-webkit.NoUncountedMemberChecker
    clang-analyzer-webkit.RefCntblBaseVirtualDtor
    clang-analyzer-webkit.UncountedLambdaCapturesChecker
    llvm-header-guard
    modernize-use-nullptr
    modernize-use-using
.....
```

**Comment by danielbisar** [comment]
The reason seems to be that clang-tidy always includes some checks:

```
---
Checks: >
  llvm-header-guard,
  modernize-use-nullptr,
  modernize-use-using,
WarningsAsErrors: ""
AnalyzeTemporaryDtors: false
FormatStyle: none
---

```

~/ws/src/project $ clang-tidy-15 -p ../../build --config-file .clang-tidy -list-checks ./subfolder/code.cpp
---
Checks:          "clang-diagnostic-*,clang-analyzer-*,llvm-header-guard, modernize-use-nullptr, modernize-use-using,\n"
WarningsAsErrors: ''
HeaderFilterRegex: ''
AnalyzeTemporaryDtors: false
FormatStyle:     none
CheckOptions:
  - key:             llvm-else-after-return.WarnOnConditionVariables
    value:           'false'
  - key:             modernize-loop-convert.MinConfidence
    value:           reasonable
  - key:             modernize-replace-auto-ptr.IncludeStyle
    value:           llvm
  - key:             cert-str34-c.DiagnoseSignedUnsignedCharComparisons
    value:           'false'
  - key:             google-readability-namespace-comments.ShortNamespaceLines
    value:           '10'
  - key:             cert-err33-c.CheckedFunctions
    value:           '::aligned_alloc;::asctime_s;::at_quick_exit;::atexit;::bsearch;::bsearch_s;::btowc;::c16rtomb;::c32rtomb;::calloc;::clock;::cnd_broadcast;::cnd_init;::cnd_signal;::cnd_timedwait;::cnd_wait;::ctime_s;::fclose;::fflush;::fgetc;::fgetpos;::fgets;::fgetwc;::fopen;::fopen_s;::fprintf;::fprintf_s;::fputc;::fputs;::fputwc;::fputws;::fread;::freopen;::freopen_s;::fscanf;::fscanf_s;::fseek;::fsetpos;::ftell;::fwprintf;::fwprintf_s;::fwrite;::fwscanf;::fwscanf_s;::getc;::getchar;::getenv;::getenv_s;::gets_s;::getwc;::getwchar;::gmtime;::gmtime_s;::localtime;::localtime_s;::malloc;::mbrtoc16;::mbrtoc32;::mbsrtowcs;::mbsrtowcs_s;::mbstowcs;::mbstowcs_s;::memchr;::mktime;::mtx_init;::mtx_lock;::mtx_timedlock;::mtx_trylock;::mtx_unlock;::printf_s;::putc;::putwc;::raise;::realloc;::remove;::rename;::scanf;::scanf_s;::setlocale;::setvbuf;::signal;::snprintf;::snprintf_s;::sprintf;::sprintf_s;::sscanf;::sscanf_s;::strchr;::strerror_s;::strftime;::strpbrk;::strrchr;::strstr;::strtod;::strtof;::strtoimax;::strtok;::strtok_s;::strtol;::strtold;::strtoll;::strtoul;::strtoull;::strtoumax;::strxfrm;::swprintf;::swprintf_s;::swscanf;::swscanf_s;::thrd_create;::thrd_detach;::thrd_join;::thrd_sleep;::time;::timespec_get;::tmpfile;::tmpfile_s;::tmpnam;::tmpnam_s;::tss_create;::tss_get;::tss_set;::ungetc;::ungetwc;::vfprintf;::vfprintf_s;::vfscanf;::vfscanf_s;::vfwprintf;::vfwprintf_s;::vfwscanf;::vfwscanf_s;::vprintf_s;::vscanf;::vscanf_s;::vsnprintf;::vsnprintf_s;::vsprintf;::vsprintf_s;::vsscanf;::vsscanf_s;::vswprintf;::vswprintf_s;::vswscanf;::vswscanf_s;::vwprintf_s;::vwscanf;::vwscanf_s;::wcrtomb;::wcschr;::wcsftime;::wcspbrk;::wcsrchr;::wcsrtombs;::wcsrtombs_s;::wcsstr;::wcstod;::wcstof;::wcstoimax;::wcstok;::wcstok_s;::wcstol;::wcstold;::wcstoll;::wcstombs;::wcstombs_s;::wcstoul;::wcstoull;::wcstoumax;::wcsxfrm;::wctob;::wctrans;::wctype;::wmemchr;::wprintf_s;::wscanf;::wscanf_s;'
  - key:             cert-oop54-cpp.WarnOnlyIfThisHasSuspiciousField
    value:           'false'
  - key:             cert-dcl16-c.NewSuffixes
    value:           'L;LL;LU;LLU'
  - key:             google-readability-braces-around-statements.ShortStatementLines
    value:           '1'
  - key:             cppcoreguidelines-non-private-member-variables-in-classes.IgnoreClassesWithAllMemberVariablesBeingPublic
    value:           'true'
  - key:             google-readability-namespace-comments.SpacesBeforeComments
    value:           '2'
  - key:             modernize-loop-convert.MaxCopySize
    value:           '16'
  - key:             modernize-pass-by-value.IncludeStyle
    value:           llvm
  - key:             modernize-use-using.IgnoreMacros
    value:           'true'
  - key:             modernize-use-nullptr.NullMacros
    value:           'NULL'
  - key:             llvm-qualified-auto.AddConstToQualified
    value:           'false'
  - key:             llvm-header-guard.HeaderFileExtensions
    value:           ';h;hh;hpp;hxx'
  - key:             modernize-loop-convert.NamingStyle
    value:           CamelCase
  - key:             llvm-else-after-return.WarnOnUnfixable
    value:           'false'
  - key:             google-readability-function-size.StatementThreshold
    value:           '800'
...

So the config is read and `clang-diagnostic-*,clang-analyzer-*` seems to be always included.

**Comment by AaronBallman** [comment]
Clang diagnostics can never be disabled, but I'm surprised that clang static analyzer is enabled by default as well. @martong @Szelethus @steakhal @haoNoQ -- do any of you know if this is expected behavior for clang-tidy?

**Comment by steakhal** [comment]
> I'm surprised that clang static analyzer is enabled by default as well

Yup, this is a problem. We have encountered this in the past.

The situation is quite bad and there are multiple symptoms/problems:
 - Only `core` analyzer checkers are intended to be enabled 'by default' along with a  couple of others, but `alpha.*` and `debug.*` packages are **not** intended to be enabled unless you know what you are doing.
 - There are other checkers which may require analyzer config options to be defined to a certain value. The only such option is `aggressive-binary-operation-simplification=true` which is required by `alpha.cplusplus.IteratorModeling,alpha.cplusplus.STLAlgorithmModeling,debug.DebugContainerModeling,debug.DebugIteratorModeling` checkers and if you enabled any of these checkers and the option is not `true`, then it will report an error message from which you cannot figure out what checker required this option, effectively sabotaging you from running the analyzer/clang-tidy.
 - If someone has downstream checks/checkers under a 'new' package, that will be also enabled unconditionally. So it's especially annoying for us :D

We mitigated this by depending on the `CodeChecker` explicitly enabling/disabling all checks. Besides that we had to apply a couple of patches for adding `-checks=-*` to all the `clang-tidy` tests using the `--config-file` option.
We did not explore the root cause nor tried fixing the underlying issue due to tight schedule, but we did know about this.
